content/Improve your Rails App searches with Elasticsearch - Leandro Dominguez.md

```diff
@@ -0,0 +1,213 @@
+
+## Let’s get down to work!
+
+### What are we going to build?
+
+We’re gonna build a really simple microblogging searcher, where we are going to be available to find posts accepting possible typos and highlighting the matching results really fast and performant. For this, we will use text analyzers and tokenizers ([natural language processing tools]([https://en.wikipedia.org/wiki/Natural_language_processing](https://en.wikipedia.org/wiki/Natural_language_processing))) that come out the box with Elastic.
+
+**Step 1: Including required gems**
+
+Add to your Rails Gemfile the two Elastic gems:
+
+```ruby
+gem 'elasticsearch-model', '~> 7.1.0'
+gem 'elasticsearch-rails', '~> 7.1.0'
+```
+
+and install them running **bundle install**.
+
+**Step 2: Adding Elastic behaviour to our model**
+
+Given a simplified `Post` model with only a `title`, `body` and a `topic`, include the next lines into it:
+
+```ruby
+class Post < ApplicationRecord
+ include Elasticsearch::Model
+ include Elasticsearch::Model::Callbacks
+end
+```
+
+The first `include` will let us make Elasticsearch calls in a friendly ActiveRecord-like syntax. This way we can, for example, just call `Post.search(query)` for searching. Otherwise we should manually make an API call to the Elastic server instance every time we want to interact with our data.
+
+The second include adds callbacks in our model, so every time we create, delete or edit a `Post`, this action will be replicated into Elasticsearch data.
+
+This last point is very important. Take into account we have two independent databases: one in which we’re gonna persist all the model data as you’re used to (Postgres) and this other new one for making fast searches, suggestions and use all other Elasticsearch fantastic features.
+
+Sometimes (for performance or even business requirements), you don’t want to update data in Elasticsearch instantaneously every time a record has been changed in our main database (this is customizable but we won’t tackle this in the current article).
+
+**Step 3: Defining our Elasticsearch index**
+
+In Elasticsearch we don’t have tables or rows. Our models are stored in ‘_indexes’_ and each instance is called ‘_document’_. Although Elastic can automatically detect the attributes types of our model, we are going to declare them explicitly so we have more control over them by adding the following to your model:
+
+```ruby
+ settings do
+   mappings dynamic: false do
+     indexes :id, index: :not_analyzed
+     indexes :title, type: :text, analyzer: :english
+     indexes :body, type: :text, analyzer: :english
+     indexes :topic, type: :keyword
+   end
+ end
+```
+
+What are we doing here?
```

Comment:
I think you may also want to explain the `dynamic: false` part, I was missing that when reading

Reply:
It was in the first version but since I thought was too large I started to take things out. But I don't this makes any difference. I'm adding it.